src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/eventretry/ClosureRetryService.kt

```diff
@@ -34,4 +41,37 @@ class ClosureRetryService(
     TransactionClosureRetriedEvent(transactionId.value(), transactionRetriedData)
 
   override fun newTransactionStatus(): TransactionStatusDto = TransactionStatusDto.CLOSURE_ERROR
+
+  override fun validateRetryEventVisibilityTimeout(
+    baseTransaction: BaseTransaction,
+    visibilityTimeout: Duration
+  ): Boolean {
+    val paymentTokenValidityOffset = Duration.ofSeconds(paymentTokenValidityTimeOffset.toLong())
+    val paymentTokenDuration = getPaymentTokenDuration(baseTransaction)
+    val paymentTokenValidityEnd =
+      baseTransaction.creationDate.plus(paymentTokenDuration).toInstant()
+    val retryEventVisibilityInstant = Instant.now().plus(visibilityTimeout)
+    // Performing check against payment token validity end and retry event visibility timeout
+    // A configurable paymentTokenValidityOffset is added to retry event in order to avoid sending
+    // retry event too
+    // closer to the payment token end
+    val paymentTokenStillValidAtRetry =
+      paymentTokenValidityEnd.minus(paymentTokenValidityOffset).isAfter(retryEventVisibilityInstant)
+    if (!paymentTokenStillValidAtRetry) {
+      logger.warn(
+        "No closure retry event send for transaction with id: [${baseTransaction.transactionId.value()}]. Retry event visibility timeout: [$retryEventVisibilityInstant], will be after payment token validity end: [$paymentTokenValidityEnd] with offset: [$paymentTokenValidityOffset]")
+    }
+    return paymentTokenStillValidAtRetry
+  }
+
+  private fun getPaymentTokenDuration(baseTransaction: BaseTransaction): Duration =
+    when (baseTransaction) {
+      is BaseTransactionWithPaymentToken ->
+        Duration.ofSeconds(
+          baseTransaction.transactionActivatedData.paymentTokenValiditySeconds.toLong())
+      is BaseTransactionWithClosureError ->
+        getPaymentTokenDuration(baseTransaction.transactionAtPreviousState)
```

Comment:
The second when case is used to extract, recursively, the payment token duration from the transactionAtPreviousState contained into the BaseTransactionWithClosureError domain object. BaseTransactionWithClosureError extends BaseTransaction directly so it's because this check is needed here.

I opened the following PR on commons library to fix BaseTransactionWithClosureError hierarchy https://github.com/pagopa/pagopa-ecommerce-commons/pull/85

Reply:
After applying the modifications to commons library (see PR above) this is fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/89/commits/dc28f956746e77984c574d52432d0b84388f3c73